group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -190,6 +216,118 @@ boolean canTransitionFrom(CoordinatorState state) {
         abstract boolean canTransitionFrom(CoordinatorState state);
     }
 
+    /**
+     * Implements the CoordinatorTimer interface. This class keeps track of all the
+     * schedule timers for a coordinator/partition.
+     *
+     * When a timer is cancelled or overridden, the previous timer is guaranteed to
+     * not be executed even if it already expired and got push to the event processor.
+     *
+     * When a timer fails with an unexpected exception, the timer is reschedule with
+     * a backoff.
+     */
+    class InternalCoordinatorTimer implements CoordinatorTimer<U> {
+        /**
+         * The logger.
+         */
+        final Logger log;
+
+        /**
+         * The topic partition.
+         */
+        final TopicPartition tp;
+
+        /**
+         * The scheduled timers keyed by their key.
+         */
+        final Map<String, TimerTask> tasks = new HashMap<>();
```

Comment:
do we need any sort of locking on this map?

Reply:
The event processor guarantees that only one operations for a given partition is executed at the time so we should be fine here.

Reply:
I guess I wasn't sure about adding to the map and running the event. We may be only doing one operation that adds to the map at a time, but does that also mean we will only hit the time to do the event during this time?

Reply:
I guess I also realize I don't fully understand why we need to have a timer for these events. But maybe I need to re-read the KIP.

Reply:
Scheduling a timeout is only done from within the execution of an event.

For the usages, you can take a look at the next PR: https://github.com/apache/kafka/pull/13963. In short, we need timers to implement session and revocation timeouts.

Reply:
Hmmm --- maybe I'm very confused now. Is the timeout actually a timeout -- ie if we haven't done the event by a given time then we return with error? I think I missed that and though we were waiting for the timer to hit a time to execute the event.

Reply:
No. The timer schedules an event when the timer expires. The event executes the timeout operation. If the timeout operation raises an error, the error is used to complete the future of the event. This is how the write event works.

Reply:
Ok -- that very much confused me so maybe we can add comments about how this. 😓 
I will need to reread a lot of this 😓

Reply:
Updated the javadoc. Let me know if it makes things clearer.

Reply:
Thanks. I also misunderstood Timer and had to read its javadoc. 
> turns timeouts of a regular {@link Timer} into {@link CoordinatorWriteEvent} events which are executed by the {@link CoordinatorEventProcessor}

With respect to the above line, it is fair to understand this as a Timer that is used to track when an operation should time out must be modified to handle the CoordinatorWriteEvents respects the threading model of the coordinator runtime. This modified Timer (CoordinatorTimer) is associated with the event via the TimerTask.

Then this task will be added and will handle timeouts appropriately now?

Reply:
Ah I see your other comment in the method. 👍